File: theseus_gui/src/helpers/skin_manager.js
import { invoke } from '@tauri-apps/api/tauri'
import { tauri } from '@tauri-apps/api'
import { SkinViewer } from 'skinview3d'
import { handleError } from '@/store/state.js'
import { ref } from 'vue'

// Returns SkinCache for user from cache
export async function get_user_skin_data(id) {
  return await invoke('plugin:skin|skin_get_user_skin_data', { id })
}

// Sets users skin
export async function set_skin(skin, arms, user) {
  return await invoke('plugin:skin|skin_set_skin', { skin, arms, user })
}

// Sets users cape
export async function set_cape(capeid, token) {
  return await invoke('plugin:skin|skin_set_cape', { capeid, token })
}

// Returns true if image is 64x64
export async function check_image(path) {
  return await invoke('plugin:skin|skin_check_image', { path })
}

// Returns true if skin is not in library
export async function check_skin(skin, id) {
  return await invoke('plugin:skin|skin_check_skin', { skin, id })
}

// Returns cape info
export async function get_cape_data(cape, key) {
  return await invoke('plugin:skin|skin_get_cape_data', { cape, key })
}

// Stores all users skin data to cache
export async function cache_users_skins() {
  return await invoke('plugin:skin|skin_cache_users_skins')
}

// Caches new user on login
export async function cache_new_user_skin(user) {
  return await invoke('plugin:skin|skin_cache_new_user_skin', { user })
}

// Saves SkinCache to the manager
export async function save_skin(user, data, name, model, skinid) {
  return await invoke('plugin:skin|skin_save_skin', { user, data, name, model, skinid })
}

// Updates the skin save
export async function update_skins(saves) {
  return await invoke('plugin:skin|skin_update_skins', { saves })
}

// Gets all saved Skins
export async function get_skins() {
  return await invoke('plugin:skin|skin_get_skins')
}

export async function get_heads() {
  account_heads.value = await invoke('plugin:skin|skin_get_heads')
}

export async function get_launcher_names(path, installer) {
  return await invoke('plugin:skin|skin_get_launcher_names', { path, installer })
}

export async function import_skin(id, path, installer) {
  return await invoke('plugin:skin|skin_import_skin', { id, path, installer })
}

export async function get_render(skinData) {
  let arms = skinData.arms
  if (arms == 'classic') arms = 'default'
  const skinViewer = new SkinViewer({
    width: 144,
    height: 144,
    preserveDrawingBuffer: true,
  })
  skinViewer.camera.rotation.x = -0.62
  skinViewer.camera.rotation.y = 0.534
  skinViewer.camera.rotation.z = 0.348
  skinViewer.camera.position.x = -20
  skinViewer.camera.position.y = 10
  skinViewer.camera.position.z = 38

  let skin = skinData.skin
  if (!skin.startsWith('data:image/png;base64,')) skin = tauri.convertFileSrc(skinData.skin)
  await skinViewer.loadSkin(skin, { model: arms })
  const cape = await get_cape_data(skinData.cape, 'url').catch(handleError)
  if (cape !== 'no cape') {
    await skinViewer.loadCape(cape)
  }

  skinViewer.render()
  // delay is required, if not, sometimes the skin isnt loaded before the canvas is saved to img
  await sleep(50)
  const image = skinViewer.canvas.toDataURL()
  skinViewer.dispose()
  return image
}

export const account_heads = ref({})

export const loaded_skins = ref(false)

const sleep = (ms = 0) => new Promise((resolve) => setTimeout(resolve, ms))

File: theseus_gui/src-tauri/src/api/skin_manager.rs
use std::{collections::HashMap, path::PathBuf};

use crate::api::Result;
use theseus::{skin_manager::{self, SkinSave, SkinCache}, auth::Credentials};

use uuid::Uuid;

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("skin")
        .invoke_handler(tauri::generate_handler![
            skin_get_user_skin_data,
            skin_get_heads,
            skin_set_skin,
            skin_delete_skin,
            skin_import_skin,
            skin_check_image,
            skin_get_cape_data,
            skin_cache_users_skins,
            skin_cache_new_user_skin,
            skin_get_mojang_launcher_path,
            skin_get_mojang_launcher_names,
            skin_save_skin,
            skin_get_skins,
            skin_set_cape
        ])
        .build()
}

// Get image size
// invoke('plugin:skin|check_image',path)
#[tauri::command]
pub async fn skin_check_image(path: String) -> Result<bool> {
    Ok(skin_manager::check_image(path).await?)
}

// Sets player's skin
// invoke('plugin:skin|skin_set_skin', { skin, arms, user })
#[tauri::command]
pub async fn skin_set_skin(skin: String, arms: String, user: Credentials) -> Result<bool> {
    Ok(skin_manager::set_skin(skin, arms, user).await?)
}

// Sets the players cape
// invoke('plugin:skin|skin_set_cape', { capeid, token })
#[tauri::command]
pub async fn skin_set_cape(capeid: String, token: String) -> Result<bool> {
    Ok(skin_manager::set_cape(capeid, token).await?)
}

// Returns cape info
// invoke('plugin:skin|skin_set_cape', { capeid, token })
#[tauri::command]
pub async fn skin_get_cape_data(cape: String, key: String) -> Result<String> {
    Ok(skin_manager::get_cape_data(cape, key).await?)
}

// Gets the current skin, cape, and updates the unlocked cape list
// invoke('plugin:skin|skin_get_player_info', { user, token })
#[tauri::command]
pub async fn skin_get_user_skin_data(id: Uuid) -> Result<SkinCache> {
    Ok(skin_manager::get_user_skin_data(id).await?)
}

// 
// 
#[tauri::command]
pub async fn skin_cache_users_skins() -> Result<bool> {
    Ok(skin_manager::cache_users_skins().await?)
}

// 
// 
#[tauri::command]
pub async fn skin_cache_new_user_skin(user: Credentials) -> Result<bool> {
    Ok(skin_manager::cache_new_user_skin(user).await?)
}

#[tauri::command]
pub async fn skin_save_skin(user: Uuid, data: SkinCache, name: String, model: String, skinid: String) -> Result<bool> {
    Ok(skin_manager::save_skin(user, data, name, model, skinid).await?)
}

#[tauri::command]
pub async fn skin_delete_skin(id: Uuid) -> Result<bool> {
    Ok(skin_manager::delete_skin(id).await?)
}

#[tauri::command]
pub async fn skin_get_heads() -> Result<HashMap<Uuid, String>> {
    Ok(skin_manager::get_heads().await?)
}

#[tauri::command]
pub async fn skin_get_skins() -> Result<Vec<SkinSave>> {
    Ok(skin_manager::get_skins().await?)
}

#[tauri::command]
pub async fn skin_get_mojang_launcher_path() -> Result<PathBuf> {
    Ok(skin_manager::get_mojang_launcher_path().await?)
}

#[tauri::command]
pub async fn skin_get_mojang_launcher_names(path: PathBuf) -> Result<Vec<skin_manager::MojangNames>> {
    Ok(skin_manager::get_mojang_launcher_names(path).await?)
}

#[tauri::command]
pub async fn skin_import_skin(name: String, path: PathBuf, user: Uuid) -> Result<()> {
    Ok(skin_manager::import_skin(name, path, user).await?)
}